src/bbs2gh/Services/BbsArchiveDownloader.cs

```diff
@@ -0,0 +1,90 @@
+using System;
+using System.IO;
+using Renci.SshNet;
+
+namespace OctoshiftCLI.BbsToGithub.Services;
+
+public class BbsArchiveDownloader : IDisposable
+{
+    private const int DOWNLOAD_PROGRESS_REPORT_INTERVAL_IN_SECONDS = 10;
+    private const string DEFAULT_BBS_SHARED_HOME_DIRECTORY = "/var/atlassian/application-data/bitbucket/shared";
+    private const string EXPORT_ARCHIVE_SOURCE_DIRECTORY = "data/migration/export";
+
+    private readonly ScpClient _scpClient;
+    private readonly OctoLogger _log;
+    private DateTime _nextProgressReport;
+
+    internal FileSystemProvider FileSystemProvider = new();
+    internal Action<string, FileInfo> ScpDownload;
```

Comment:
can we not mock ScpClient itself? instead of doing this indirection via function pointers

Reply:
without mocking it'd be impossible to unit test it. I had two choices, either to create an abstraction (`IScpClient` or something, that is going to be just and adapter) to be mocked later in tests or the delegate approach that I already took.
That being said if you think it is too much, I can get rid of the unit tests and make the downloader so much simpler.

Reply:
Why can't we do `new Mock<ScpClient>()` ?

Reply:
oh I think I see, it's because the functions aren't virtual in ScpClient. Interestingly if we were to use SftpClient it has an ISftpClient we could mock.

Reply:
addressed in 8c3a31526e42e969e554ca6f7fcbd5db878b2665